File: code/python/729.py
# https://leetcode.com/problems/my-calendar-i/
from bisect import bisect_left
_MAX_VALUE_ = 10 ** 9 + 1

class MyCalendar2:
    def __init__(self):
        self.store = [(-1, -1), (_MAX_VALUE_, _MAX_VALUE_)]

    def book(self, start: int, end: int) -> bool:
        idx = bisect_left(self.store, (start, end))
        if end > self.store[idx][0]:
            return False
        if start < self.store[idx - 1][1]:
            return False
        self.store.insert(idx, (start, end))
        return True

File: code/cpp/729.cpp
// https://leetcode.com/problems/my-calendar-i/
class MyCalendar {
private:
	const int _MAX_VALUE_ = 1e9 + 1;
	set<pair<int, int>> store;

public:
	MyCalendar() {
		store.insert({ -1,-1 });
		store.insert({ _MAX_VALUE_,_MAX_VALUE_ });
	}

	bool book(int start, int end) {
		const pair<int, int> event{ start, end };
		const auto nextEvent = store.lower_bound(event);
		if (nextEvent->first < end)
			return false;
		const auto prevEvent = prev(nextEvent);
		if (prevEvent->second > start)
			return false;
		store.insert(event);
		return true;
	}
};
